File: arc/go/types/channels.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package types

// Chan returns a channel type wrapping the given value type.
func Chan(valueType Type) Type {
	return Type{Kind: KindChan, Elem: &valueType}
}

// ReadChan returns a channel type annotated for read access.
func ReadChan(valueType Type) Type {
	return Type{Kind: KindChan, Elem: &valueType, ChanDirection: ChanDirectionRead}
}

// WriteChan returns a channel type annotated for write access.
func WriteChan(valueType Type) Type {
	return Type{Kind: KindChan, Elem: &valueType, ChanDirection: ChanDirectionWrite}
}

// NewChannels creates a new Channels with empty read and write sets.
func NewChannels() Channels {
	return Channels{Read: make(map[uint32]string), Write: make(map[uint32]string)}
}

File: console/src/feature/mqtt/task/channels.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type channel, NotFoundError, type Synnax as Client } from "@synnaxlabs/client";
import { DataType, errors } from "@synnaxlabs/x";

export const retrieveChannel = async (
  client: Client,
  key: channel.Key,
): Promise<channel.Channel | null> => {
  try {
    return await client.channels.retrieve(key);
  } catch (e) {
    if (NotFoundError.matches(e)) return null;
    throw errors.fromUnknown(e);
  }
};

export const channelExists = async (
  client: Client,
  key: channel.Key,
): Promise<boolean> => (await retrieveChannel(client, key)) != null;

/**
 * Creates a channel for values of dataType. A fixed-density channel gets its own
 * index named `${name}_time`; a variable-density one is virtual.
 */
export const createChannel = async (
  client: Client,
  name: string,
  dataType: string,
): Promise<channel.Channel> => {
  if (new DataType(dataType).isVariable)
    return await client.channels.create({ name, dataType, virtual: true });
  const index = await client.channels.create({
    name: `${name}_time`,
    dataType: "timestamp",
    isIndex: true,
  });
  return await client.channels.create({ name, dataType, index: index.key });
};
